### ISSUE TITLE
Masking not working - Keras lstm model.predict seems to padd random values at the wrong timesteps

### ISSUE THREAD

**Issue by mastersg7** [keras-team/keras#12380]
I have a lstm model that predict time series. Before propagation the inputs to the hidden layers, the features pass through a masking layer that masks observations for rows where all columns contain the value 0. All the observation squences have variable length, afterwhich are padded to obtain the same length eg. [1 2 3 4 0 0], [8 5 0 0 0 0]

Wanting to verify my performance metric during validation and testing, I overfit the training data and provide the training data as validation_data to the model. Similarly, I provide the training data to model.predict(). During training the validation errors goes very low, as it should. However after predicting in testing and using the same performance metric, the error is higher. Looking at the training data target and the predicted values. I can see a strange pattern. mainly that the model predict well for the orginal sequence except the last one but then padds the rest of the sequence with a certain random number. for instance [0.8 2 3 -1.222 -1.222 -1.222] It seems to me that model.predict() doesnt capture well how many zero paddings there are, meaning it start padding one timestep too early. As a results the performance measure is way off compared to the validation measure.

Is there a difference between the predict function used during validation and model.predict()? If there is a difference. is it possible to call the predict function used during validation? Furthermore Any ideas on how to solve this? 

Please find script with some sample data [here](https://github.com/mastersg7/FPL). Verify by looking at yhat and y_train

**Comment by thelahunginjeet** [comment]
I'm running into a similar issue that seems to be related to whether you use GPUs when you predict (keras 2.3.1, tf 2.1.0) -  meaning, I get this predict problem if I'm using GPUs but not if I only use the CPU. Is anyone looking into this?
